### PR TITLE
Extend multi style to finite-sized particles

### PR DESCRIPTION
**Pull request by jtclemm**
**Summary**

This pull request includes new NPair classes to extend support of multi neighboring to finite sized particles. This allows efficient simulations of granular systems with large polydispersity.

**Related Issue(s)**

None

**Author(s)**

Ishan Srivastava (LBL)
Joel Clemmer (SNL) - jtclemm@sandia.gov

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

No foreseen issues

**Implementation Notes**

Only added new classes, no changes to other LAMMPS files or features. Npair classes are based on pre-existing multi style NPair classes. The primary difference is the use of atom->radius to determine whether particles are in contact.

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

None

### CODE REVIEW

**Comment by akohlmey** [comment]
@jtclemm thanks for the contribution. Have you checked if this is compatible with `KOKKOS` and `USER-OMP?
Since those also have granular pair styles.

**Comment by jtclemm** [comment]
This PR will not work with OMP. Looking at the USER-OMP source code, my understanding is that I'd just need to create OMP equivalents for these NPair classes to add compatibility. Considering versions of npair multi and size already exist, this shouldn't be difficult. Let me see if I can get it working.

I have not checked Kokkos but will test it. I'm unfamiliar with Kokkos development so I generally don't know what is required for code to be compatible.

**Comment by akohlmey** [comment]
> This PR will not work with OMP. Looking at the USER-OMP source code, my understanding is that I'd just need to create OMP equivalents for these NPair classes to add compatibility. Considering versions of npair multi and size already exist, this shouldn't be difficult. Let me see if I can get it working.

Yes. I don't think that there was much benefit from multi-threading the granular neighbor list builds, so they are aliases. Ping me, if you need help with that, but I would appreciate it if you could sort it out on your own. I am currently carrying a fairly heavy load with preparing the upcoming stable release which is much more than usual since we have a whole extra bunch of new stuff and significant changes to code infrastructure and particularly the documentation. Checking all the corner cases is time consuming.
> 
> I have not checked Kokkos but will test it. I'm unfamiliar with Kokkos development so I generally don't know what is required for code to be compatible.

**Comment by akohlmey** [comment]
> I have not checked Kokkos but will test it. I'm unfamiliar with Kokkos development so I generally don't know what is required for code to be compatible.

You probably need help from @stanmoore1 for that. Most of us do.

**Comment by jtclemm** [comment]
I added files to support USER-OMP. I was unfamiliar with some of the notation so I lifted it straight from other pre-existing npair OMP classes. The new classes compile and run with OMP on my system. Hopefully I didn't overlook anything!

I checked Kokkos but saw that Kokkos only supports bin-style neighbor lists. Adding a new comm/neighbor style to Kokkos is a bit beyond my current abilities so I think this PR will have to remain incompatible with Kokkos.

**Comment by stanmoore1** [comment]
Yeah Kokkos doesn't yet support NP_MULTI, so I think we can leave it as-is wrt to Kokkos.

File: doc/src/comm_modify.rst
Comment (sjplimp) [inline doc/src/comm_modify.rst:93]: I think the neighbor doc page also needs a similar small addition to the mutli explanation, to say how it treats particles with finite size (e.g. granular).

**Review by sjplimp — APPROVED** [review]

**Comment by sjplimp** [comment]
@akohlmey after doc change, I think is good to merge.  Sorry, I clicked approve after asking for the doc change :(

File: doc/src/comm_modify.rst
Comment (jtclemm) [reply doc/src/comm_modify.rst:93]: Good idea, I added a similar comment to the neighbor documentation clarifying how cutoffs are defined for finite sized particles.

**Pull request merged**
